### PR TITLE
2020-01-05 Patches

### PR DESCRIPTION
**Pull request by mdblack98**

### CODE REVIEW

File: include/hamlib/rig.h [hunk 1]
```diff
@@ -93,24 +93,6 @@
 #endif
 
 #include <time.h>
```
Comment (N0NB) [inline include/hamlib/rig.h]: It seems like this line should have been moved to sleep.c as well and the local include of rig.h omitted.

**Comment by mdblack98** [comment]
Yup...that's a touch cleaner....
de Mike W9MDB

 

    On Monday, January 6, 2020, 01:28:34 AM CST, Nate Bargmann <notifications@github.com> wrote:  
 
 
@N0NB commented on this pull request.

In include/hamlib/rig.h:
> @@ -93,24 +93,6 @@
 #endif
 
 #include <time.h>

It seems like this line should have been moved to sleep.c as well and the local include of rig.h omitted.

**Comment by N0NB** [comment]
Am seeing this on my MinGW build:

```
  CC       dttsp.lo
In file included from dttsp.c:37:0:
dttsp.c: In function ‘fetch_meter’:
dttsp.c:382:37: warning: format ‘%ld’ expects argument of type ‘long int’, but argument 5 has type ‘unsigned int’ [-Wformat=]
             rig_debug(RIG_DEBUG_ERR,"%s: sizeof(float)!=4, instead = %ld\n",__func__, sizeof(float));
                                     ^
../include/hamlib/rig.h:2367:92: note: in definition of macro ‘rig_debug’
 #define rig_debug(debug_level,fmt,...) { char xxxbuf[16384];snprintf(xxxbuf,sizeof(xxxbuf),fmt,__VA_ARGS__);rig_debug(debug_level,fmt,##__VA_ARGS__); }
                                                                                            ^~~
```

**Comment by mdblack98** [comment]
Fixed it to be portable.
Mike

 

    On Tuesday, January 7, 2020, 07:51:04 AM CST, Nate Bargmann <notifications@github.com> wrote:  
 
 
Am seeing this on my MinGW build:
  CC       dttsp.lo
In file included from dttsp.c:37:0:
dttsp.c: In function ‘fetch_meter’:
dttsp.c:382:37: warning: format ‘%ld’ expects argument of type ‘long int’, but argument 5 has type ‘unsigned int’ [-Wformat=]
             rig_debug(RIG_DEBUG_ERR,"%s: sizeof(float)!=4, instead = %ld\n",__func__, sizeof(float));
                                     ^
../include/hamlib/rig.h:2367:92: note: in definition of macro ‘rig_debug’
 #define rig_debug(debug_level,fmt,...) { char xxxbuf[16384];snprintf(xxxbuf,sizeof(xxxbuf),fmt,__VA_ARGS__);rig_debug(debug_level,fmt,##__VA_ARGS__); }
                                                                                            ^~~

**Comment by N0NB** [comment]
I just pulled to match your latest and get this on MinGW:

```
  CC       termios.lo
termios.c: In function ‘win32_serial_read’:
termios.c:1632:17: warning: implicit declaration of function ‘hl_usleep’ [-Wimplicit-function-declaration]
                 hl_usleep(100);     /* don't hog the CPU while waiting */
                 ^~~~~~~~~
```

This is preventing the build of rigctl.exe and subsequent programs in the tests directory:

```
  CCLD     rigctl.exe
rigctl-rigctl.o: In function `main':
/home/nate/builds/hamlib-4.0~git/tests/rigctl.c:515: undefined reference to `hl_usleep'
collect2: error: ld returned 1 exit status
Makefile:947: recipe for target 'rigctl.exe' failed
make[1]: *** [rigctl.exe] Error 1
make[1]: Leaving directory '/home/nate/builds/hamlib-4.0~git/tests'
Makefile:605: recipe for target 'install-recursive' failed
make: *** [install-recursive] Error 1
```

**Comment by mdblack98** [comment]
Did you re-run bootstrap?configure has to regenerated to pick up the new nanosleep check.
de Mike W9MDB
 

    On Wednesday, January 8, 2020, 07:39:10 AM CST, Nate Bargmann <notifications@github.com> wrote:  
 
 
I just pulled to match your latest and get this on MinGW:
  CC       termios.lo
termios.c: In function ‘win32_serial_read’:
termios.c:1632:17: warning: implicit declaration of function ‘hl_usleep’ [-Wimplicit-function-declaration]
                 hl_usleep(100);     /* don't hog the CPU while waiting */
                 ^~~~~~~~~

This is preventing the build of rigctl.exe and subsequent programs in the tests directory:
  CCLD     rigctl.exe
rigctl-rigctl.o: In function `main':
/home/nate/builds/hamlib-4.0~git/tests/rigctl.c:515: undefined reference to `hl_usleep'
collect2: error: ld returned 1 exit status
Makefile:947: recipe for target 'rigctl.exe' failed
make[1]: *** [rigctl.exe] Error 1
make[1]: Leaving directory '/home/nate/builds/hamlib-4.0~git/tests'
Makefile:605: recipe for target 'install-recursive' failed
make: *** [install-recursive] Error 1

**Comment by N0NB** [comment]
* On 2020 08 Jan 08:07 -0600, Michael Black wrote:
> Did you re-run bootstrap?configure has to regenerated to pick up the new nanosleep check.

Yup.  I do a working tree clean with git clean and I rm -rf * my build
directory before each pull.

The MinGW builds take place from the tarball generated by make
distcheck.  My Debian desktop didn't give the warning on hl_sleep.

73, Nate

-- 

"The optimist proclaims that we live in the best of all
possible worlds.  The pessimist fears this is true."

Web: https://www.n0nb.us
Projects: https://github.com/N0NB
GPG fingerprint: 82D6 4F6B 0E67 CD41 F689 BBA6 FB2C 5130 D55A 8819

**Comment by mdblack98** [comment]
The question would be why include/hamlib/rig.h's definition of hl_usleep is not being seen.And why src/Makefile.am must not have sleep.c in it.
Sounds like you didn't get the current version or missed some of it somehow. 
Mike
    On Wednesday, January 8, 2020, 08:40:03 AM CST, Nate Bargmann <notifications@github.com> wrote:  
 
 * On 2020 08 Jan 08:07 -0600, Michael Black wrote:
> Did you re-run bootstrap?configure has to regenerated to pick up the new nanosleep check.

Yup. I do a working tree clean with git clean and I rm -rf * my build
directory before each pull.

The MinGW builds take place from the tarball generated by make
distcheck. My Debian desktop didn't give the warning on hl_sleep.

73, Nate

-- 

"The optimist proclaims that we live in the best of all
possible worlds. The pessimist fears this is true."

Web: https://www.n0nb.us
Projects: https://github.com/N0NB
GPG fingerprint: 82D6 4F6B 0E67 CD41 F689 BBA6 FB2C 5130 D55A 8819

**Comment by N0NB** [comment]
Fixing termios.c is easy by simply adding:

```#include <hamlib/rig.h>```

I'm still puzzling over why I'm still getting a linker error until I can recall how to dump the symbols from a .dll.

**Comment by mdblack98** [comment]
Is there another place for the list of files used for elk creation?

**Comment by N0NB** [comment]
I think I may have found it and put in a patch for it.
Try again....

Mike

**Comment by N0NB** [comment]
I think the issue with why I'm seeing the linker fail on MinGW is that in src/sleep.c the sleep() and usleep() functions are wrapped by `#ifdef HAVE_NANOSLEEP` and that macro is not defined in `config.h` as Windows doesn't have nanosleep nor the other two functions.  As a result none of the three functions defined in src/sleep.c show up in the symbol table of src/.libs/sleep.o.

My guess is that for Windows the function in the removed include/hl_sleep.h file or similar will need to be restored.  It sucks, but that's Windows for ya!

**Comment by N0NB** [comment]
In other words, Windows will need a custom nanosleep function provided and only compiled for Windows.

**Comment by mdblack98** [comment]
nanosleep should be defined in pthread.h
I'm hoping that including pthread.h in rig.h allows configure to pick up HAVE_NANOSLEEP on Win32
de Mike W9MDB

 

    On Wednesday, January 8, 2020, 06:45:45 PM CST, Nate Bargmann <notifications@github.com> wrote:  
 
 
In other words, Windows will need a custom nanosleep function provided and only compiled for Windows.

**Comment by N0NB** [comment]
Linking rigctl.exe still results in the following linker error in my MinGW build:

```
  CCLD     rigctl.exe
rigctl-rigctl.o: In function `main':
/home/nate/builds/hamlib-4.0~git/tests/rigctl.c:515: undefined reference to `hl_usleep'
collect2: error: ld returned 1 exit status
Makefile:947: recipe for target 'rigctl.exe' failed
make[1]: *** [rigctl.exe] Error 1
make[1]: Leaving directory '/home/nate/builds/hamlib-4.0~git/tests'
Makefile:605: recipe for target 'install-recursive' failed
make: *** [install-recursive] Error 1
```

I checked `include/config.h` again and `HAVE_NANOSLEEP` is undefined.

I ensured that I had pulled your latest commits and built from a fresh tarball.

**Comment by mdblack98** [comment]
Should be fixed now...
de Mike W9MDB

 

    On Thursday, January 9, 2020, 07:41:34 PM CST, Nate Bargmann <notifications@github.com> wrote:  
 
 
Linking rigctl.exe still results in the following linker error in my MinGW build:
  CCLD     rigctl.exe
rigctl-rigctl.o: In function `main':
/home/nate/builds/hamlib-4.0~git/tests/rigctl.c:515: undefined reference to `hl_usleep'
collect2: error: ld returned 1 exit status
Makefile:947: recipe for target 'rigctl.exe' failed
make[1]: *** [rigctl.exe] Error 1
make[1]: Leaving directory '/home/nate/builds/hamlib-4.0~git/tests'
Makefile:605: recipe for target 'install-recursive' failed
make: *** [install-recursive] Error 1

I checked include/config.h again and HAVE_NANOSLEEP is undefined.

I ensured that I had pulled your latest commits and built from a fresh tarball.

**Comment by N0NB** [comment]
It compiles after I patched `src/Makefile.am` to distribute `sleep.h`.

**Pull request merged**